### DISCUSSION TITLE
Glb file having different appearance

### DISCUSSION THREAD

**Discussion by mbalajee** [google/filament#7851] [Q&A]
Gltf viewer online | Filament
---|---
<img width="1520" alt="Screenshot 2024-05-14 at 2 26 36 PM" src="https://github.com/google/filament/assets/17508872/b4d2c2ec-44f9-4450-a166-8b11f4b32109">|<img width="1136" alt="Screenshot 2024-05-14 at 2 26 44 PM" src="https://github.com/google/filament/assets/17508872/c64ac3b5-8af6-4cfe-bcb6-4d0c306248f6">

Same gltf file appears differently on online viewer and Filament.  What changes do I have to make to get the appearance from the online viewer?  
* I have tried disabling all options in the debug view
* Removed the "materials" block from `gltf` file

Model: https://drive.google.com/drive/folders/175h337ECA1590Nk0mX5VRtnkJpL2xdHJ?usp=sharing

**Comment by romainguy** [comment]
@poweifeng Looks like we don't generate the normals properly. Didn't you fix this recently?

**Reply by poweifeng** [reply]
I did.  But you'll have to enable this path explicitly.  Let me write a response.

**Comment by poweifeng** [comment]
There is some background of this issue in #6358 .

To address your problem (specifically the gltf you're using).  You can apply the following diff to `samples/gltf_viewer.cpp`

```
diff --git a/samples/gltf_viewer.cpp b/samples/gltf_viewer.cpp
index 80a06298e..91ae2c781 100644
--- a/samples/gltf_viewer.cpp
+++ b/samples/gltf_viewer.cpp
@@ -747,7 +747,16 @@ int main(int argc, char** argv) {
                 createJitShaderProvider(engine, OPTIMIZE_MATERIALS) :
                 createUbershaderProvider(engine, UBERARCHIVE_DEFAULT_DATA, UBERARCHIVE_DEFAULT_SIZE);

-        app.assetLoader = AssetLoader::create({engine, app.materials, app.names });
+        AssetConfigurationExtended ext = {
+            .gltfPath = filename.c_str(),
+        };
+        AssetConfiguration config = {
+            .engine = engine,
+            .materials = app.materials,
+            .names = app.names,
+            .ext = &ext,
+        };
+        app.assetLoader = AssetLoader::create(config);
         app.mainCamera = &view->getCamera();
         if (filename.isEmpty()) {
             app.asset = app.assetLoader->createAsset(
```
This is the output

<img width="602" alt="image" src="https://github.com/google/filament/assets/48371888/2e55f772-4e85-4ef7-8170-fb8d4ff367d6">

The reason why this path is kind of hidden is that it has limitations such as only being able to handle local-to-disk gltfs. So it's not really ready for prime-time. To introduce the fix properly, we have a tracking issue #7782 to create a new API for gltf loading. Maybe this current workaround (adding the `ext` object) is enough for your use case?

**Reply by mbalajee** [reply]
@poweifeng Thanks for the background.  Patch fix works on `gltf_viewer.cpp` but the `AssetConfiguration` class is not available `gltfio-android` library (am on latest `main`).  We're using filament on Android to display the gltf files.  Any workaround in current state of Filament for Android ?

**Reply by poweifeng** [reply]
Unfortunately the "extended" algos are not supported on Android right now - due to the fact that storage access in Android is defined through the ResourceLoader class. The only option I can think of is to fork filament/gltfio and make a bit of changes to the AssetLoader/ResourceLoader implementations.

**Reply by poweifeng** [reply]
Or maybe use blender or related software to change your gltf into one that's explicitly flat-shaded (duplicated vertices and normals, normals associated with faces etc).

**Reply by mbalajee** [reply]
Yeah I just created a new JNI function with an additional parameter for file path (we save models in Files directory not in Assets folder) and called the `AssetLoader::create` with the new extended config and it works

```java
extern "C" JNIEXPORT jlong JNICALL
Java_com_google_android_filament_gltfio_AssetLoader_nCreateAssetLoaderExtended(JNIEnv* env, jclass,
                                                                       jlong nativeEngine, jobject provider, jlong nativeEntities, jstring filePath) {
    .....
    const char *nativeFilePath = env->GetStringUTFChars(filePath, nullptr);
    AssetConfigurationExtended ext = {
            .gltfPath = nativeFilePath
    };
    AssetConfiguration config = {
            .engine = engine,
            .materials = materialProvider,
            .names = names,
            .ext = &ext,
    };
    return (jlong) AssetLoader::create(config);
}
```

Any reason why this can't be built into the `gltfio-android` and `filament-utils-android` libraries?  Because we would like to use filament as a dependency instead of building from source so it's easier to keep Filament updated

**Reply by mbalajee** [reply]
Ah spoke too soon... model appears all white, guess I'll need to make some changes in `Utility.cpp#loadCgltfBuffers`
<img src="https://github.com/google/filament/assets/17508872/509563bb-1358-41c6-8877-3475b4ae6eac" width=30% height=50%>

**Reply by mbalajee** [reply]
I changed conditions that was excluding code path for Android platform and I put debug points on the code changes in the [fix PR](https://github.com/google/filament/pull/7776/files#diff-2801a8206180648982cb86348e95e76649d1e979f3b4eb04a6257d1d90227491) & they're all executing.  Model still appears fully in white and no shadows, changes in Indirect light & other color related properties Debug view does not affect the appearance (remains white) but if i disable View post processing it goes all black.  It appears correctly in `gltf_viewer.cpp` app.   Do I need to make any other change? 

@poweifeng

**Reply by mbalajee** [reply]
Here are some logs, that I added in AssetLoader and ResourceLoader, from running on Android,

```
FEngine (64 bits) created at 0xb40000724a020640 (threading is enabled)
Using ASurfaceTexture
FEngine resolved backend: OpenGL
[Google (Apple)], [Android Emulator OpenGL ES Translator (Apple M3 Max)], [OpenGL ES 3.0 (
Feature level: 1
Active workarounds: 
GL_MAX_TEXTURE_MAX_ANISOTROPY_EXT = 0
GL_MAX_COMBINED_TEXTURE_IMAGE_UNITS = 80
GL_MAX_DRAW_BUFFERS = 8
GL_MAX_RENDERBUFFER_SIZE = 16384
GL_MAX_SAMPLES = 4
GL_MAX_TEXTURE_IMAGE_UNITS = 16
GL_MAX_TRANSFORM_FEEDBACK_SEPARATE_ATTRIBS = 4
GL_MAX_UNIFORM_BLOCK_SIZE = 65536
GL_MAX_UNIFORM_BUFFER_BINDINGS = 80
GL_NUM_PROGRAM_BINARY_FORMATS = 0
GL_UNIFORM_BUFFER_OFFSET_ALIGNMENT = 256
OS version: 34
Backend feature level: 1
FEngine feature level: 1
Extended Configuration supported
config ext path /data/user/0/com.google.android.filament.gltf/files/models/21_KB.glb
Texture Decoder has 1 background threads.
FFilamentAsset constructor true
createPrimitives primitiveCount 1
Output prim Vertices count 108
Output prim Indices count 108
createPrimitives primitiveCount 1
Output prim Vertices count 108
Output prim Indices count 108
createPrimitives primitiveCount 1
Output prim Vertices count 108
Output prim Indices count 108
createPrimitives primitiveCount 1
Output prim Vertices count 108
Output prim Indices count 108
createPrimitives primitiveCount 1
Output prim Vertices count 108
Output prim Indices count 108
isUsingExtendedAlgo true
Requested linear format but KTX contains a sRGB format. 
Requested linear format but KTX contains a sRGB format. 
```

Logs when running on Mac (looks similar to android),
```
FEngine (64 bits) created at 0x14486e400 (threading is enabled)
FEngine resolved backend: Metal
Selected physical device 'Apple M3 Max'
Supported GPU families: 
  MTLGPUFamilyCommon3
  MTLGPUFamilyApple7
  MTLGPUFamilyMac2
Features:
  readWriteTextureSupport: true
Backend feature level: 3
FEngine feature level: 3
Requested linear format but KTX contains a sRGB format. 
Requested linear format but KTX contains a sRGB format. 
Extended Configuration supported
config ext path /Users/balajimurali/Downloads/21_KB_new.glb
Content size 4964 File /Users/balajimurali/Downloads/21_KB_new.glb
Creating asset for file /Users/balajimurali/Downloads/21_KB_new.glb
FFilamentAsset constructor true
createPrimitives primitiveCount 1
Output prim Vertices count 108
Output prim Indices count 108
createPrimitives primitiveCount 1
Output prim Vertices count 108
Output prim Indices count 108
createPrimitives primitiveCount 1
Output prim Vertices count 108
Output prim Indices count 108
createPrimitives primitiveCount 1
Output prim Vertices count 108
Output prim Indices count 108
createPrimitives primitiveCount 1
Output prim Vertices count 108
Output prim Indices count 108
File parsed
Node count 14Texture Decoder has 14 background threads.
isUsingExtendedAlgo true
```

**Reply by mbalajee** [reply]
Issue fixed by,
https://github.com/google/filament/pull/7917
